liquibase/liquibase
# 1054 add not null constraint causes data too long exception

## Environment
**Liquibase Version**: `4.5.0`

**Liquibase Integration & Version**: Spring Boot / Maven in my case but not specific to this integration

**Liquibase Extension(s) & Version**:  N/A

**Database Vendor & Version**: MySQL

**Operating System Type & Version**:  Not OS specific but I'm running Fedora 34

## Pull Request Type


* \[x] Bug fix (non-breaking change which fixes an issue.)
* \[ ] Enhancement/New feature (non-breaking change which adds functionality)
* \[ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description
Fix #1054

A clear and concise description of the issue being addressed.  Additional guidance [here](https://liquibase.jira.com/wiki/spaces/LB/pages/1274904896/How+to+Contribute+Code+to+Liquibase+Core).

* Describe the actual problematic behavior.
* Ensure private information is redacted.

## Steps To Reproduce
See:

* [https://github.com/liquibase/liquibase/issues/1054#issue-587536866](https://github.com/liquibase/liquibase/issues/1054#issue-587536866)
* [https://github.com/liquibase/liquibase/issues/1054#issuecomment-804873212](https://github.com/liquibase/liquibase/issues/1054#issuecomment-804873212) (comment)

## Actual Behavior
Liquibase generates invalid SQL when setting a default value to a boolean column.

Taken from [https://github.com/liquibase/liquibase/issues/1054#issue-587536866:](https://github.com/liquibase/liquibase/issues/1054#issue-587536866:)

```java
<changeSet id="XXX" author="XXX">
    <addNotNullConstraint tableName="XX" columnName="XX" defaultNullValue="1" columnDataType="BIT(1)"/>
</changeSet>
```

Generates the following error:

```java
Error setting up or running Liquibase: Migration failed for change set xxx.xml::xxxx::XXX:
[ERROR]      Reason: liquibase.exception.DatabaseException: (conn=28) Data too long for column
```

## Expected/Desired Behavior
Liquibase generates a valid SQL update statement in the form of:

```java
UPDATE FOO SET BAR = false WHERE BAR IS NULL
```

## Fast Track PR Acceptance Checklist:


<

!--- If you're unsure about any of these, just ask us in a comment. We're here to help|width=200,height=183!

-->

* \[x] Build is successful and all new and existing tests pass
* \[x] Added [Unit Test(s)](https://liquibase.jira.com/wiki/spaces/LB/pages/1274937609/How+to+Write+Liquibase+Core+Unit+Tests)
* \[ ] Added [Integration Test(s)](https://liquibase.jira.com/wiki/spaces/LB/pages/1276608569/How+to+Write+Liquibase+Core+Integration+Tests)
* \[ ] Added [Test Harness Test(s)](https://github.com/liquibase/liquibase-test-harness/pulls)
* \[ ] Documentation Updated

## Need Help?
Come chat with us on our [discord channel](https://discord.com/channels/700506481111597066/700506481572839505)

- - -
## Dev Handoff Notes (Internal Use)
#### Links
* Fixed Issue: [https://github.com/liquibase/liquibase/issues/1054](https://github.com/liquibase/liquibase/issues/1054)
* Test Results: [https://github.com/liquibase/liquibase/pull/2071/checks](https://github.com/liquibase/liquibase/pull/2071/checks)

#### Testing
* Steps to Reproduce: [https://github.com/liquibase/liquibase/issues/1054](https://github.com/liquibase/liquibase/issues/1054)
* Guidance:
  * Impact: only impacts `<addNotNullConstraint>` when dataType is a boolean like field for that database. But the logic is applied for all database types

#### Dev Verification
I used a changelog like this:

```java
<changeSet id="1" author="XXX">
        <createTable tableName="test_table">
            <column name="id" type="int"/>
            <column name="col_name_bit" type="BIT(1)"/>
            <column name="col_name_boolean" type="BOOLEAN"/>
        </createTable>
    </changeSet>

    <changeSet id="2" author="XXX">
        <insert tableName="test_table">
            <column name="id" valueNumeric="1"/>
            <column name="col_name_bit" valueNumeric="0"/>
            <column name="col_name_boolean" valueBoolean="false"/>
        </insert>
        <insert tableName="test_table">
            <column name="id" valueNumeric="2"/>
        </insert>
    </changeSet>

    <changeSet id="test-bit" author="XXX">
        <addNotNullConstraint tableName="test_table" columnName="col_name_bit" defaultNullValue="1" columnDataType="BIT(1)"/>
    </changeSet>

    <changeSet id="test-bool" author="XXX">
        <addNotNullConstraint tableName="test_table" columnName="col_name_boolean" defaultNullValue="true" columnDataType="BOOLEAN"/>
    </changeSet>
```

With the fix, the update correctly runs.

## Test Requirements (Liquibase Internal QA)
The manual testing is only required for **MySQL** database.

**Setup**

Execute update command against MySQL database (changelog file is attached):

`liquibase update --changelog-file lb2187-changelog.xml --labels create,fill`

Verify that table `test_table` was successfully created and has 2 rows with data as shown below

!Screenshot from 2021-12-15 18-43-12.png|width=25%!

**Manual Tests**

_Verify that_ `update-sql` _generates correct sql code for_ `bit` _type._

* execute `liquibase update-sql --changelog-file lb2187-changelog.xml --labels bit`
* expected sql code: 

```java
UPDATE [db_name].test_table SET col_name_bit = 1 WHERE col_name_bit IS NULL;
ALTER TABLE [db_name].test_table MODIFY col_name_bit BIT(1) NOT NULL;
```

_Verify that_ `update` _is successful for_ `bit` _type_

* execute `liquibase update --changelog-file lb2187-changelog.xml --labels bit` 
* update is successful
* no exception is thrown
* `col_name_bit` value for row with `id=2` is now **1** instead of **[NULL]**

_Verify that_ `update-sql` _generates correct sql code for_ `boolean` _type._

* execute `liquibase update-sql --changelog-file lb2187-changelog.xml --labels bool`
* expected sql code: 

```java
UPDATE [db_name].test_table SET col_name_boolean = true WHERE col_name_boolean IS NULL;
ALTER TABLE [db_name].test_table MODIFY col_name_boolean BIT(1) NOT NULL;
```

_Verify that_ `update` _is successful for_ `boolean` _type_

* execute `liquibase update --changelog-file lb2187-changelog.xml --labels bool`
* update is successful
* no exception is thrown
* `col_name_boolean` value for row with `id=2` is now **1** instead of **[NULL]**

**Automated Tests**

* A new test-harness test is required to verify fix behavior against other supported database types.



┆Issue is synchronized with this [Jira Bug](https://datical.atlassian.net/browse/LB-2187) by [Unito](https://www.unito.io)
